app/src/main/java/ch/hikemate/app/model/profile/ProfileRepositoryFirestore.kt

```diff
@@ -0,0 +1,129 @@
+package ch.hikemate.app.model.profile
+
+import android.util.Log
+import com.google.android.gms.tasks.Task
+import com.google.firebase.Firebase
+import com.google.firebase.auth.auth
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+
+/**
+ * A Firestore implementation of the profile repository.
+ *
+ * @property db The Firestore database.
+ */
+class ProfileRepositoryFirestore(val db: FirebaseFirestore) : ProfileRepository {
+
+  private val collectionPath = "profiles"
+
+  override fun getNewUid(): String {
+    return db.collection(collectionPath).document().id
+  }
+
+  override fun init(onSuccess: () -> Unit) {
+    // Check if the user is logged in
+    // If the user is logged in, call onSuccess
+    Firebase.auth.addAuthStateListener {
+      if (it.currentUser != null) {
+        onSuccess()
+      }
+    }
+  }
+
+  override fun getProfileById(
+      id: String,
+      onSuccess: (Profile) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    // Check that the profile exists
+
+    db.collection(collectionPath).document(id).get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        val profile = task.result?.let { documentToProfile(it) }
+        if (profile != null) {
+          onSuccess(profile)
+        } else {
+          onFailure(Exception("Profile not found"))
+        }
+      } else {
+        task.exception?.let { e ->
+          Log.e("ProfileRepositoryFirestore", "Error getting document", e)
+          onFailure(e)
+        }
+      }
+    }
+  }
+
+  override fun addProfile(profile: Profile, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performFirestoreOperation(
+        db.collection(collectionPath).document(profile.id).set(profile), onSuccess, onFailure)
+  }
```

Comment:
I took this from the bootcamp code.
I think the reason behind this is to use `getNewUid` in the view/viewmodel as we may need it. If we create a newUid inside of the repository, It would be a headache to get it in the viewmodel after calling `addProfile`.

Do you think we should do it another way ?

Reply:
From what you are saying it makes sense, using `getNewUid` in the view model, but what if an existing ID is used instead?

The `addProfile` method should look more like something such as

```kotlin
override fun addProfile(profile: Profile, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
  db.collection(collectionPath).document(profile.id).get().addOnCompleteListener { task ->
    if (task.isSuccessful) {
      val document = task.result
      if (document != null && document.exists()) {
        onFailure(Exception("Profile with ID ${profile.id} already exists"))
      } else {
        updateProfile(profile, onSuccess, onFailure)
      }
    } else {
      task.exception?.let { e ->
        Log.e("ProfileRepositoryFirestore", "Error checking if profile exists", e)
        onFailure(e)
      }
    }
  }
}
```

1. See if the profile ID exists
2. If it does not, use the same logic as `updateProfile`

Reply:
I agree. The only thing I changed from what you proposed is to use `getProfileById` in order to check if the profile exists. Like this:
```kotlin
  override fun addProfile(profile: Profile, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
    getProfileById(
        profile.id,
        onSuccess = {
          // If we found a profile, we can't create a new one with the same id
          onFailure(Exception("Profile with id ${profile.id} already exists"))
        },
        onFailure = {
          // If we didn't find a profile, we can create a new one
          // Call updateProfile to create the profile as it is the same implementation
          updateProfile(profile, onSuccess, onFailure)
        })
  }
```